### PR TITLE
Fall Motion

### PR DESCRIPTION
**Pull request by schmidma**
## Introduced Changes

This PR introduces a fall motion to protect the robot when falling.
For this, the new `FallState::Sitting` is added

## ToDo / Known Issues

- [x] if the robot pulls back its arms when avoiding obstacles, they might stuck behind the back in the forward falling motion.

## How to Test

push the robot, let it stumble, let it fall

### CODE REVIEW

**Comment by okiwi6** [comment]
Fixes #543

**Review by schluis — CHANGES_REQUESTED** [review]
Looks very good, thank you for your work! Sometimes the robot gets stuck, lays down on its arms and does not continue falling. I think we should implement something like a timeout before merging.

File: crates/control/src/fall_state_estimation.rs [hunk 1]
```diff
@@ -1,41 +1,50 @@
...
 
 #[context]
 pub struct CycleContext {
-    backward_gravitational_difference: AdditionalOutput<f32, "backward_gravitational_difference">,
+    fallen_up_gravitational_difference: AdditionalOutput<f32, "backward_gravitational_difference">,
     filtered_angular_velocity: AdditionalOutput<Vector3<f32>, "filtered_angular_velocity">,
     filtered_linear_acceleration: AdditionalOutput<Vector3<f32>, "filtered_linear_acceleration">,
     filtered_roll_pitch: AdditionalOutput<Vector2<f32>, "filtered_roll_pitch">,
-    forward_gravitational_difference: AdditionalOutput<f32, "forward_gravitational_difference">,
+    fallen_down_gravitational_difference: AdditionalOutput<f32, "forward_gravitational_difference">,
```
[the remark this answers is not in the document]
Comment (schmidma) [reply crates/control/src/fall_state_estimation.rs]: Done

File: crates/control/src/fall_state_estimation.rs [hunk 2]
```diff
@@ -49,44 +58,29 @@ impl FallStateEstimation {
...
-        let inertial_measurement_unit = convert_to_right_handed_coordinate_system(
-            context.sensor_data.inertial_measurement_unit,
-        );
-
-        let robot_to_inertial_measurement_unit = Isometry3::from_parts(
-            Translation3::identity(),
-            UnitQuaternion::from_axis_angle(&Vector3::y_axis(), PI),
-        );
-        let inertial_measurement_unit_to_robot = robot_to_inertial_measurement_unit.inverse();
+        let inertial_measurement_unit = context.sensor_data.inertial_measurement_unit;
```
[the remark this answers is not in the document]
Comment (schmidma) [reply crates/control/src/fall_state_estimation.rs:77]: I actually cannot tell. I did not understand what the translation was doing, it was incorrect, in that it produced incorrect imu values and it is not necessary. The *raw* sensor values are in the correct frame already

File: crates/control/src/fall_state_estimation.rs [hunk 3]
```diff
@@ -98,107 +92,106 @@ impl FallStateEstimation {
...
+                start_time: context.cycle_time.start_time,
+            },
+            (FallState::Upright, Some(_), Some(facing)) => FallState::Fallen { facing },
+            (current @ FallState::Falling { start_time, .. }, None, None) => {
+                if context
+                    .cycle_time
+                    .start_time
+                    .duration_since(start_time)
+                    .unwrap()
+                    > *context.falling_timeout
```
[the remark this answers is not in the document]
Comment (schmidma) [reply crates/control/src/fall_state_estimation.rs:164]: This happens e.g. when the robot is picked up while falling or similar... usually only with manual interaction.

File: crates/control/src/motion/fall_protector.rs [hunk 4]
```diff
@@ -57,179 +52,127 @@ pub struct MainOutputs {
...
+        Ok(Self::default())
     }
 
     pub fn cycle(&mut self, context: CycleContext) -> Result<MainOutputs> {
-        let current_positions = context.sensor_data.positions;
-        let mut head_stiffness = 1.0;
-
-        self.roll_pitch_filter
-            .update(context.sensor_data.inertial_measurement_unit.roll_pitch);
+        let measured_positions = context.sensor_data.positions;
```
[the remark this answers is not in the document]
Comment (schmidma) [reply crates/control/src/motion/fall_protector.rs:60]: Which *other instances*? Can you please mark them?

File: crates/control/src/motion/fall_protector.rs [hunk 5]
```diff
@@ -57,179 +52,127 @@ pub struct MainOutputs {
...
+            FallState::Upright | FallState::Fallen { .. } => {
+                context.motion_safe_exits[MotionType::FallProtection] = true;
+                return Ok(MainOutputs::default());
+            }
+            FallState::Falling {
+                start_time,
+                direction,
+            } => (start_time, direction),
+        };
         context.motion_safe_exits[MotionType::FallProtection] = false;
```
[the remark this answers is not in the document]
Comment (schmidma) [reply crates/control/src/motion/fall_protector.rs:72]: hmm... :thinking: not sure. This way independently of any decision and adjustment of the `match` statement above, it is guaranteed that the safe exit is set to false before we continue

File: crates/control/src/motion/fall_protector.rs [hunk 6]
```diff
@@ -1,51 +1,46 @@
...
 
-#[derive(Deserialize, Serialize)]
-pub struct FallProtector {
-    start_time: SystemTime,
-    interpolator: MotionInterpolator<Joints<f32>>,
-    roll_pitch_filter: LowPassFilter<Vector2<f32>>,
-    last_fall_state: FallState,
-    fallen_time: Option<SystemTime>,
+#[derive(Clone, Copy, Debug)]
+enum Phase {
```
[the remark this answers is not in the document]
Comment (schmidma) [reply crates/control/src/motion/fall_protector.rs:17]: Isn't that clear as this type is private in the `fall_protector`? This seems to be a violation of *do-not-repeat-yourself*

File: crates/control/src/motion/fall_protector.rs [hunk 7]
```diff
@@ -57,179 +52,127 @@ pub struct MainOutputs {
...
-                    .duration_since(fallen_start)
-                    .expect("time ran backwards")
-                    >= context.fall_protection.time_prolong_ground_impact =>
-            {
-                context.motion_safe_exits[MotionType::FallProtection] = true;
-                self.fallen_time = None;
-            }
-            _ => (),
-        }
+        let joints_command = MotorCommands {
```
[the remark this answers is not in the document]
Comment (schmidma) [reply crates/control/src/motion/fall_protector.rs:133]: yep, done

File: crates/control/src/fall_state_estimation.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (schluis) [reply crates/control/src/fall_state_estimation.rs:77]: okay xD

File: crates/control/src/fall_state_estimation.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (schluis) [reply crates/control/src/fall_state_estimation.rs:164]: Okay, makes sense

File: crates/control/src/motion/fall_protector.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (schluis) [reply crates/control/src/motion/fall_protector.rs:72]: Okay, agree

File: crates/control/src/motion/fall_protector.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (schluis) [reply crates/control/src/motion/fall_protector.rs:17]: Yes, but it is not the phase of the `FallProtector` but the `Fall`. I just remember being confused because I was unsure what this `Phase` actually refers to.

**Comment by schluis** [comment]
Still need to implement `FallState::Gorilla`

File: crates/types/src/fall_state.rs [hunk 8]
```diff
@@ -1,17 +1,39 @@
...
 use serialize_hierarchy::SerializeHierarchy;
 
-use crate::motion_command::{Facing, FallDirection};
+#[derive(Clone, Copy, Debug, Serialize, Deserialize, SerializeHierarchy)]
+pub enum Side {
+    Left,
+    Right,
+}
 
-#[derive(Clone, Copy, Debug, Serialize, Deserialize, SerializeHierarchy, PartialEq)]
```
Comment (larskna) [inline crates/types/src/fall_state.rs:6]: Why did you remove PartialEq ?

File: etc/motions/stand_up_back.json [hunk 9]
```diff
@@ -3,167 +3,44 @@
...
-      "elbow_roll": 0.0,
-      "wrist_yaw": 0.0,
-      "hand": 0.0
+      "elbow_roll": 1.376039981842041,
+      "elbow_yaw": -0.14883995056152344,
+      "hand": 0,
+      "shoulder_pitch": 1.9620280265808105,
+      "shoulder_roll": -0.26695799827575684,
+      "wrist_yaw": 0.056715965270996094
     },
```
Comment (larskna) [inline etc/motions/stand_up_back.json:23]: Why did you change the order?

File: etc/motions/stand_up_front.json [hunk 10]
```diff
@@ -5,36 +5,36 @@
...
-      "hip_pitch": 0,
-      "knee_pitch": 0,
-      "ankle_pitch": 1.0471975511965976,
-      "ankle_roll": 0
+      "ankle_pitch": -0.9970581531524658,
+      "ankle_roll": -0.0030260086059570312,
+      "hip_pitch": 0.3128941059112549,
+      "hip_roll": 0.015382051467895508,
+      "hip_yaw_pitch": -0.01222991943359375,
+      "knee_pitch": 0.07827591896057129
```
Comment (larskna) [inline etc/motions/stand_up_front.json:37]: Why did you change the order ?

**Review by larskna — CHANGES_REQUESTED** [review]
We looked at the code. 
The test on the robots looked greate, while the hardware worked.

File: etc/motions/stand_up_back.json (hunk 9 above)
Comment (schluis) [reply etc/motions/stand_up_back.json:23]: This is directly copied from Twix and is in alphabetical order

Comment (larskna) [reply etc/motions/stand_up_back.json:23]: okay

File: crates/control/src/motion/stand_up_sitting.rs [hunk 11]
```diff
@@ -0,0 +1,79 @@
...
+        let condition_input = context.condition_input;
+
+        context.motion_safe_exits[MotionType::StandUpSitting] = false;
+
+        self.interpolator
+            .advance_by(last_cycle_duration, condition_input);
+
+        if self.interpolator.is_finished() {
+            context.motion_safe_exits[MotionType::StandUpSitting] = true;
+        }
```
[the remark this answers is not in the document]
Comment (schluis) [reply crates/control/src/motion/stand_up_sitting.rs:62]: Done

File: crates/control/src/fall_state_estimation.rs [hunk 12]
```diff
@@ -99,107 +94,132 @@ impl FallStateEstimation {
...
+                {
+                    FallState::Sitting {
+                        start_time: context.cycle_time.start_time,
+                    }
+                } else {
+                    current
+                }
+            }
+            (FallState::Fallen { .. }, None, None) => FallState::Upright,
+            (FallState::Fallen { .. }, None | Some(_), Some(facing)) => {
```
[the remark this answers is not in the document]
Comment (schluis) [reply crates/control/src/fall_state_estimation.rs]: Done

File: crates/control/src/fall_state_estimation.rs [hunk 13]
```diff
@@ -99,107 +94,132 @@ impl FallStateEstimation {
...
+                {
+                    FallState::Upright
+                } else {
+                    current
+                }
+            }
+            (
+                current @ FallState::Falling { start_time, .. }
+                | current @ FallState::Sitting { start_time },
+                None | Some(_),
```
[the remark this answers is not in the document]
Comment (schluis) [reply crates/control/src/fall_state_estimation.rs]: Done

File: crates/control/src/foot_bumper_filter.rs [hunk 14]
```diff
@@ -134,7 +134,7 @@ impl FootBumperFilter {
 
         self.check_for_bumper_errors(&context);
 
-        if *fall_state != FallState::Upright {
+        if matches!(fall_state, FallState::Upright) {
```
[the remark this answers is not in the document]
Comment (schluis) [reply crates/control/src/foot_bumper_filter.rs:136]: Reverted

File: crates/types/src/fall_state.rs (hunk 8 above)
Comment (schluis) [reply crates/types/src/fall_state.rs:6]: @schmidma has forgotten why. Reverted

File: crates/control/src/motion/fall_protector.rs [hunk 15]
```diff
@@ -57,179 +54,129 @@ pub struct MainOutputs {
...
+            }
+            (FallDirection::Backward { side: Side::Right }, FallPhase::Late) => *context.back_late,
+        };
 
-        let body_stiffnesses = if self.roll_pitch_filter.state().y.abs()
-            > context.fall_protection.ground_impact_angular_threshold
-        {
-            BodyJoints::fill(context.fall_protection.ground_impact_body_stiffness)
+        let is_head_protected = measured_positions.head.pitch.abs() < *context.reached_threshold
+            && measured_positions.head.yaw.abs() < *context.reached_threshold;
```
[the remark this answers is not in the document]
Comment (schluis) [reply crates/control/src/motion/fall_protector.rs]: Done

File: crates/control/src/fall_state_estimation.rs [hunk 16]
```diff
@@ -1,42 +1,52 @@
...
+use nalgebra::{vector, Rotation3, Vector2, Vector3};
 use serde::{Deserialize, Serialize};
 
 use context_attribute::context;
 use filtering::low_pass_filter::LowPassFilter;
 use framework::{AdditionalOutput, MainOutput};
 use types::{
-    fall_state::FallState,
-    motion_command::{Facing, FallDirection},
-    parameters::FallStateEstimationParameters,
```
[the remark this answers is not in the document]
Comment (schluis) [reply crates/control/src/fall_state_estimation.rs:13]: Done

File: crates/control/src/motion/fall_protector.rs [hunk 17]
```diff
@@ -1,51 +1,48 @@
...
-    joints::{body::BodyJoints, head::HeadJoints, Joints},
-    motion_command::{FallDirection, MotionCommand},
-    motion_selection::{MotionSafeExits, MotionSelection, MotionType},
+    fall_state::{FallDirection, FallState, Side},
+    joints::{
+        arm::ArmJoints, body::BodyJoints, head::HeadJoints, leg::LegJoints, mirror::Mirror, Joints,
+    },
+    motion_selection::{MotionSafeExits, MotionType},
     motor_commands::MotorCommands,
-    parameters::{FallProtectionParameters, FallStateEstimationParameters},
```
[the remark this answers is not in the document]
Comment (schluis) [reply crates/control/src/motion/fall_protector.rs:20]: Done

File: crates/control/src/fall_state_estimation.rs [hunk 18]
```diff
@@ -99,107 +94,132 @@ impl FallStateEstimation {
...
-            UnitQuaternion::from_axis_angle(&Vector3::y_axis(), -FRAC_PI_2),
-        );
-        let robot_to_fallen_up = Isometry3::from_parts(
-            Translation3::identity(),
-            UnitQuaternion::from_axis_angle(&Vector3::y_axis(), FRAC_PI_2),
-        );
-
-        let fallen_direction = if (self.linear_acceleration_filter.state()
+        let robot_to_fallen_down = Rotation3::from_axis_angle(&Vector3::y_axis(), -FRAC_PI_2);
+        let robot_to_fallen_up = Rotation3::from_axis_angle(&Vector3::y_axis(), FRAC_PI_2);
```
[the remark this answers is not in the document]
Comment (schluis) [reply crates/control/src/fall_state_estimation.rs]: Done

File: crates/types/src/sensor_data.rs [hunk 19]
```diff
@@ -7,9 +8,9 @@ use crate::joints::Joints;
 #[derive(Clone, Copy, Debug, Default, Deserialize, Serialize, SerializeHierarchy)]
 pub struct InertialMeasurementUnitData {
     // Linear acceleration is coming from a left handed coordinate system
-    pub linear_acceleration: Vector3<f32>,
-    pub angular_velocity: Vector3<f32>,
-    pub roll_pitch: Vector2<f32>,
+    pub linear_acceleration: Vector3<Robot>,
+    pub angular_velocity: Vector3<Robot>,
```
Comment (schmidma) [inline crates/types/src/sensor_data.rs:12]: I am not sure whether this is correct and/or we want to do that. What do you think @knoellle ?

Comment (knoellle) [reply crates/types/src/sensor_data.rs:12]: I think if you want to use a vector here, it should be framed accordingly.
IIRC nalgebra just uses tuples for euler angles, but of course that doesn't implement things like addition, uniform s.aling, etc.
I'm fine with this solution.

File: etc/parameters/head.P0000074A09S28J00025.json [hunk 20]
```diff
@@ -14,5 +14,10 @@
         -1.9930126667022705
       ]
     }
-  }
+  },
+  "fall_state_estimation": {
+    "difference_to_sitting_threshold": 1.5,
+    "gravitational_acceleration_threshold": 7.0
+  },
+  "player_number": "Five"
```
Comment (schmidma) [inline etc/parameters/head.P0000074A09S28J00025.json:22]: accidental?

File: crates/control/src/fall_state_estimation.rs [hunk 21]
```diff
@@ -1,42 +1,66 @@
...
+    filtered_roll_pitch: AdditionalOutput<Vector2<Robot>, "filtered_roll_pitch">,
+    fallen_down_gravitational_difference:
+        AdditionalOutput<f32, "fallen_down_gravitational_difference">,
+    fallen_sitting_gravitational_difference:
+        AdditionalOutput<f32, "fallen_standing_gravitational_difference">,
+    fallen_up_gravitational_difference: AdditionalOutput<f32, "fallen_up_gravitational_difference">,
+    upright_gravitational_difference: AdditionalOutput<f32, "upright_gravitational_difference">,
+    difference_to_sitting: AdditionalOutput<f32, "difference_to_sitting">,
+    falling_direction: AdditionalOutput<Option<Direction>, "falling_direction">,
+    fallen_direction: AdditionalOutput<Option<Kind>, "fallen_direction">,
```
Comment (schmidma) [inline crates/control/src/fall_state_estimation.rs:48]: these are all in the cycler's global scope. To reduce duplication, you can also come up with a contextual struct grouping all these outputs

**Review by schmidma — COMMENTED** [review]
very minor things

File: etc/parameters/head.P0000074A09S28J00025.json (hunk 20 above)
Comment (larskna) [reply etc/parameters/head.P0000074A09S28J00025.json:22]: deleted it 
set it in default.js

**Review by ThagonDuarte — APPROVED** [review]
